### PR TITLE
fix(steps): bound Step 7 planning and Step 9 dreaming loop lengths (#2214)

### LINKED ISSUE #2214: fix(steps): bound Step 7 planning and Step 9 dreaming loop lengths before jnp.arange/lax.scan
## Category

**Fix** — repairs a compute-safety defect that can silently consume unbounded wall-clock/memory at JAX trace time. No benchmark number is claimed by this issue; the outcome is a measurement environment that fails closed instead of hanging or OOMing.

## Summary

`alberta_framework/steps/step7.py` and `alberta_framework/steps/step9.py` validate their inner planning/dreaming loop lengths only up to `_INT32_MAX`, then hand those values directly to `jnp.arange(...)` feeding `jax.lax.scan` / `jax.vmap`. A typo'd or hostile config (for example `planning_steps=10**9`) therefore passes every existing validator and then attempts a multi-gigabyte `arange` materialization plus a billion-iteration compiled scan at trace time — an OOM or indefinite hang with no actionable error, instead of the clean host-side `ValueError` this repository's own convention requires.

The repository already has the exact machinery and precedent for rejecting this class before tracing: `alberta_framework/_scan_resources.py` (`ScanBudget`, `require_scan_steps`), applied by `core/dreaming.py` (`_DREAM_ROLLOUT_BUDGET`, `maximum_steps=10_000`), `core/temporal_context.py`, `core/compositional_features.py`, `core/feature_discovery.py`, `core/interaction_features.py`, and the Step 8 smoke entry (`_STEP8_SMOKE_BUDGET`). Steps 7 and 9 import none of it. Recent merged fixes of the same class (#2126 history-feature cap, #2128 option-count cap, #2116 feature-dim reject, option-search caps) establish that this fail-closed boundary is intended to be complete.

## Affected sites

All five values are accepted up to `2**31 - 1` by their config validators and then used as static scan lengths:

- `alberta_framework/steps/step7.py`
  - config validation accepts `planning_steps` up to `_INT32_MAX` (validator around line 278), then `jnp.arange(config.planning_steps, dtype=jnp.int32)` feeds the per-update planning `lax.scan` (line ~1221).
  - `planning_rollout_depth` accepted up to `_INT32_MAX` (around line 284), then `jnp.arange(config.planning_rollout_depth, dtype=jnp.int32)` feeds the rollout `lax.scan` inside every planning iteration (line ~1136).
- `alberta_framework/steps/step9.py`
  - `dream_candidate_count` accepted up to `_INT32_MAX` (around line 438), then `jnp.arange(config.dream_candidate_count, dtype=jnp.int32)` drives `jax.vmap(candidate_step)` (line ~775).
  - `dream_rollout_horizon` accepted up to `_INT32_MAX` (around line 431), then `jnp.arange(config.dream_rollout_horizon, dtype=jnp.int32)` feeds the dream-rollout `lax.scan` (line ~863).
  - `planning_budget` accepted up to `_INT32_MAX` (around line 428), then `jnp.arange(config.planning_budget, dtype=jnp.int32)` feeds the outer dream `lax.scan` (line ~911).

Reproduction sketch (hostile config, no GPU needed):

```python
cfg = Step7DynaConfig(planning_steps=10**9)   # passes current validation
# first step7_update trace: jnp.arange(10**9, int32) => 4 GiB alloc + 1e9-iteration scan
```

## Why this matters here

Steps 5–12 are library-surface kernels consumed by downstream integration work. A single bad config value converts into minutes-to-hours of wasted compute or a killed worker before any error surfaces, which is exactly the failure mode the existing `ScanBudget` guards were merged to eliminate elsewhere. The guard belongs at the same boundary (config validation, before `arange`), not at call sites scattered through the update functions.

## Proposed fix

1. Declare named, auditable `ScanBudget`s for Step 7 planning and Step 9 dreaming (following the `core/dreaming.py` pattern; a `maximum_steps` of 10_000 matches the existing public last-fit precedent and every known caller uses values ≤ 3).
2. Route the five config fields through `require_scan_steps` in each module's config validator so oversized values are rejected with a clear message before any array exists.
3. Failing-test-first: extend the existing `tests/test_step7_hostile_validation.py` and `tests/test_step9_hostile_validation.py` with cases proving rejection happens at config construction/validation time, plus a case proving values at the boundary are accepted.
4. No numerical change whatsoever: every previously-valid config below the caps takes the identical code path; the guard is pure addition.

## Acceptance criteria

- [ ] Oversized values for all five fields raise `ValueError` at config validation, before any `jnp.arange`.
- [ ] Existing suite passes unchanged (`pytest tests -q -o addopts=""` on affected files), ruff clean, mypy strict clean.
- [ ] Defaults and all currently-used values (≤ 3 across the tree) unaffected.
- [ ] No changes to hash-registered evidence sources, pinned artifacts, thresholds, or validators.

## Non-goals

- No new user-facing configuration knob; budgets are fixed module constants like the existing precedents.
- No change to `run_step8_scan`'s data-length-bounded scan (its length comes from caller-supplied arrays that must already exist in memory) and no change to smoke entries that already use `require_scan_steps`.
- No performance claim and no benchmark result.

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2214.

## Summary
In `alberta_framework/steps/step7.py` and `alberta_framework/steps/step9.py`, inner planning and dreaming loop lengths were bounded only up to `_INT32_MAX` before feeding static `jnp.arange(...)` scan generators. Hostile or oversized configurations could attempt multi-gigabyte memory materializations and billion-iteration compiled loops at JAX trace time.

## Proposed Changes
1. Added auditable `ScanBudget` bounds (maximum 10,000 steps matching module precedents in `core.dreaming`) for:
   - Step 7 `planning_steps` and `planning_rollout_depth`.
   - Step 9 `planning_budget`, `dream_rollout_horizon`, and `dream_candidate_count`.
2. Updated config validators in `step7.py` and `step9.py` to enforce these scan bounds during configuration validation.
3. Added unit tests in `tests/test_step7_hostile_validation.py` and `tests/test_step9_hostile_validation.py`.

## Test Plan
- `uv run pytest tests/test_step7_hostile_validation.py tests/test_step9_hostile_validation.py` (46 passed)
- `uv run ruff check alberta_framework/steps/step7.py alberta_framework/steps/step9.py tests/test_step7_hostile_validation.py tests/test_step9_hostile_validation.py` (clean)
- `uv run mypy alberta_framework/steps/step7.py alberta_framework/steps/step9.py` (clean)

### CODE REVIEW

**Review by OutstandingRetard — APPROVED** [review]
Scoped review of #2803 (fix(steps): bound Step 7 planning and Step 9 dreaming loop lengths (#2214)).
Touches 4 file(s): `alberta_framework/steps/step7.py`, `alberta_framework/steps/step9.py`, `tests/test_step7_hostile_validation.py`, `tests/test_step9_hostile_validation.py`. Lead change `alberta_framework/steps/step7.py` (+8/-3).
Matches the stated fix and is ready for maintainer merge.
No blocking concerns.

**Review by agent-cortex — APPROVED** [review]
## Independent review — APPROVE (with sandbox reproduction)

I reviewed PR #2803 against issue #2214 and independently reproduced the pre-fix vs post-fix behavior in an isolated Python 3.12 + JAX 0.7.1 CPU sandbox. The change is correct, minimal, scoped to the two files #2214 names, matches the repo's own `ScanBudget`/`core.dreaming` precedent (`maximum_steps=10_000`), and its tests pass. Along the way I found that the fix is *more necessary* for two of the five values than the issue implies — details below.

### Scope check
Diff touches only `alberta_framework/steps/step7.py`, `alberta_framework/steps/step9.py`, and the two `tests/test_step{7,9}_hostile_validation.py` files. The five caps target exactly the values #2214 enumerates. The `_require_int` change preserves the old `"must be at most int32 max"` message for existing `_INT32_MAX` callers and only emits `"must be <= {maximum}"` for the new bounds, so no other validator's contract changes.

### Design note (correct)
The PR reuses the existing `_require_int(..., maximum=budget.maximum_steps)` rather than the repo's `require_scan_steps` helper. This is the right call: `require_scan_steps` enforces `1 <= value`, but `planning_steps`/`planning_budget` legitimately allow `0` (disabled). Using `_require_int` preserves `minimum=0` for those two and `minimum=1` for the rollout/candidate fields.

### Reproduction (sandbox, network-denied): base vs head, hostile `10**9`
Base tree = the two files at `c9aba7b` overlaid on the rest of the package; HEAD = `4bab040`.

```
===================== BASE (pre-PR) =====================
  step7.planning_steps=10**9        -> REJECTED: derived Step 7 planning output bytes must fit signed int32
  step7.planning_rollout_depth=10**9-> REJECTED: derived Step 7 planning output bytes must fit signed int32
  step9.planning_budget=10**9       -> REJECTED: derived Step 9 dream work per real step must be at most 4096
  step9.dream_rollout_horizon=10**9 -> ACCEPTED (no host-side bound before JAX trace)
  step9.dream_candidate_count=10**9 -> ACCEPTED (no host-side bound before JAX trace)
===================== HEAD (PR #2803) =====================
  step7.planning_steps=10**9        -> REJECTED: planning_steps must be <= 10000
  step7.planning_rollout_depth=10**9-> REJECTED: planning_rollout_depth must be <= 10000
  step9.planning_budget=10**9       -> REJECTED: planning_budget must be <= 10000
  step9.dream_rollout_horizon=10**9 -> REJECTED: dream_rollout_horizon must be <= 10000
  step9.dream_candidate_count=10**9 -> REJECTED: dream_candidate_count must be <= 10000
```

Key finding: three of the five values (`planning_steps`, `planning_rollout_depth`, `planning_budget`) were *already* caught on base by later aggregate checks (derived-bytes / dream-work-per-real-step), so the PR mainly makes those fail earlier and with a clearer, uniform message. But **`dream_rollout_horizon` and `dream_candidate_count` were genuinely unbounded on base** — the PR closes two real fail-open holes. I confirmed both directly feed trace-time allocation:
- `step9.py:870` `jnp.arange(config.dream_rollout_horizon, dtype=jnp.int32)` → `jax.lax.scan` (`:867`). `10**9` ⇒ ~4 GB int32 `arange` + billion-iteration compiled scan at trace time.
- `step9.py:742` `jr.split(candidate_key, config.dream_candidate_count)` (~16 GB of PRNG keys) and `step9.py:782` `jnp.arange(config.dream_candidate_count, ...)`.

So the fix is not merely cosmetic hardening — for these two fields it prevents an actual OOM/hang, exactly the class #2214 targets.

### Focused tests on the real head
```
$ python3.12 -m pytest tests/test_step7_hostile_validation.py tests/test_step9_hostile_validation.py -q
46 passed in 12.89s
$ pytest ... -k scan_budget
2 passed, 44 deselected
```
The new tests exercise rejection at 10001 and acceptance at the 10000 boundary for all five fields (the step9 boundary cases correctly set `planning_budget=0` so the dream-work product check doesn't mask the per-field bound).

### Non-blocking observations
- The `10_000` per-field ceiling is looser than step9's existing `_MAX_DREAM_WORK_PER_REAL_STEP = 4096` product bound; this is fine — the product check still governs *combined* dream work, while the new per-field caps govern standalone scan lengths (e.g. `dream_rollout_horizon` when `planning_budget=0`). Coherent, defense-in-depth.
- Minor accuracy note for the record: #2214's statement that "all five values are accepted up to 2**31-1" is imprecise (only two truly were, per the base run above). This does not diminish the fix.

No correctness, security, or contract gaps found. Verified: static audit of both files and their scan consumption sites, base-vs-head behavioral reproduction, and the touched test files on the real head.

---
AI provider/model: anthropic / claude-opus-4-8
Client / agent tooling: pi-coding-agent
Contribution skill revision: elizaOS/eliza@bc84c0527374221f5c9f40ede4fbe2e62f703d8c:packages/skills/skills/contribute-to-eliza
Attribution status: self-reported
— [eliza-swarm]

**Comment by lalalune** [comment]
Reviewed: Closing after resource review: independent 10000 planning/depth limits still allow 100 million imagined steps and are not a measured resource envelope. Step 9 already has an aggregate dream-work bound. The proposed tests also omit adjusting existing Step 7 error-contract expectations. This does not provide a justified end-to-end budget.
